src/hera/workflows/script.py

```diff
@@ -606,27 +617,81 @@ def _output_annotations_used(source: Callable) -> bool:
 
 FuncIns = ParamSpec("FuncIns")  # For input types of given func to script decorator
 FuncR = TypeVar("FuncR")  # For return type of given func to script decorator
+FuncRCov = TypeVar("FuncRCov", covariant=True)
 
 ScriptIns = ParamSpec("ScriptIns")  # For input attributes of Script class
-StepIns = ParamSpec("StepIns")  # # For input attributes of Step class
-TaskIns = ParamSpec("TaskIns")  # # For input attributes of Task class
 
 
-# Pass actual classes of Script, Step and Task to bind inputs to the ParamSpecs above
+class _ScriptDecoratedFunction(Generic[FuncIns, FuncRCov], Protocol):
+    """Type assigned to functions decorated with @script."""
+
+    # Note: name is required for Step + Task, but is automatically inferred from
+    #       the name of the decorated function for @script.
+    # Note: For more details about overload-overlap, see https://github.com/python/typeshed/issues/12178
+
+    @overload
+    def __call__(  # type: ignore [overload-overlap]
+        self,
+        *,
+        name: str = ...,
+        continue_on: Optional[ContinueOn] = ...,
+        hooks: Optional[Dict[str, LifecycleHook]] = ...,
+        on_exit: Optional[Union[str, Templatable]] = ...,
+        template: Optional[Union[str, _ModelTemplate, TemplateMixin, CallableTemplateMixin]] = ...,
+        template_ref: Optional[TemplateRef] = ...,
+        inline: Optional[Union[_ModelTemplate, TemplateMixin]] = ...,
+        when: Optional[str] = ...,
+        with_sequence: Optional[_ModelSequence] = ...,
+        arguments: ArgumentsT = ...,
```

Comment:
While technically all of these are optional, I wonder if we should make `arguments` required here? At the moment, there's no way to correctly unit-test the underlying function if it takes no parameters, or it happens to have parameters with the same name as some of these. It would be easy to put `arguments={}` in your step/task if you don't want to pass arguments.

Reply:
With the change in precedence to prefer the user's function, this makes perfect sense, thanks.